src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -5,14 +5,104 @@ package freechips.rocketchip.diplomacy
 import Chisel.{defaultCompileOptions => _, _}
 import chisel3.internal.sourceinfo.{SourceInfo, UnlocatableSourceInfo}
 import chisel3.{MultiIOModule, RawModule, Reset, withClockAndReset}
-import chisel3.experimental.{ChiselAnnotation}
+import chisel3.experimental.{BaseModule, ChiselAnnotation}
+import chisel3.aop.{Select => CSelect}
 import firrtl.passes.InlineAnnotation
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.util.CompileOptions.NotStrictInferReset
 
 import scala.collection.immutable.{ListMap, SortedMap}
 import scala.util.matching._
 
+object Select {
+  type AnyMixedNode = MixedNode[_, _, _, _ <: Data, _, _, _, _ <: Data]
+
+  case class InwardEdge[Bundle <: Data, EdgeInParams](
+    params: Parameters,
+    bundle: Bundle,
+    edge: EdgeInParams,
+    node: OutwardNode[_, _, Bundle],
+  )
+
+  case class OutwardEdge[Bundle <: Data, EdgeOutParams](
+    params: Parameters,
+    bundle: Bundle,
+    edge: EdgeOutParams,
+    node: InwardNode[_, _, Bundle],
+  )
+
+  def nodes(lazyModule: LazyModule): Iterable[BaseNode] = {
+    lazyModule.nodes
+  }
+
+  private def getInwardEdges[BI <: Data, EI](node: MixedNode[_, _, EI, BI, _, _, _, _ <: Data]): Iterable[InwardEdge[BI, EI]] = {
+    node.iPorts.zip(node.in).map {
+      case ((_, node, params, _), (bundle, edge)) =>
+        InwardEdge(params, bundle, edge, node)
+    }
+  }
+
+  private def getOutwardEdges[BO <: Data, EO](node: MixedNode[_, _, _, _ <: Data, _, _, EO, BO]): Iterable[OutwardEdge[BO, EO]] = {
+    node.oPorts.zip(node.out).map {
+      case ((_, node, params, _), (bundle, edge)) =>
+        OutwardEdge(params, bundle, edge, node)
+    }
+  }
+
+  def collectInwardEdges[T](node: BaseNode)(collect: PartialFunction[InwardEdge[_ <: Data, _], T]): Iterable[T] = {
+    node match {
+      case node: AnyMixedNode =>
+        require(node.instantiated,
+          s"""Select helpers cannot be used with node "${node.name}" that has not been instantiated yet.""")
+        node.bundlesSafeNow = true
+        val result = getInwardEdges(node).collect(collect)
+        node.bundlesSafeNow = false
```

Comment:
Flipping `bundlesSafeNow` on and off every time you want to access a bundle seems like...contravening the intentionality of the flag. Either the bundles are safe to instantiate/manipulate or they aren't, seem really weird to have this code toggling it. I suspect there's duplication of functionality with your `instantiated` that should just be cleaned up/unified

Reply:
I think the reasoning behind the `bundlesSafeNow` implicitly assumed any time you accessed the bundle, it was because you wanted to assign to it. However, the Aspect model contradicts that assumption because you access bundles to annotate it, but not assign to it. I'm not arguing for any change in particular, just adding context.

Reply:
I'm open to refining things into `instantiated`(=> implies `bundlesSafeToAccessNow`) and rename the existing one to `bundlesSafeToAssignNow`, but, in the case where you are deploying an aspect that is intended to force some signals to have certain values for verification purposes, wouldn't that count as "assigning" to them?

Reply:
i thought that just reading a bundle needs to happen in a "safe" context. not assigning to it. "safe" (from my reading of the code) implied that it had a real hardware meaning at that point, for both assigning and reading.

Reply:
Sorry @mwachs5 I don't quite follow what you are saying. If "just reading...needs...safe" but "safe...implies...both assigning and reading" why do we need two flags?

Reply:
To rephrase, I'm no longer seeing the utility in this library being a nanny about keeping the user from making cross-module references through either reading from or assigning to this bundle. Seems like there are plenty of reasons to access the bundle and things that one might do with a reference to it outside the scope of its `LazyModuleImp`. Sure some of them are bad but I think that without knowing what the user is going to do with the reference we aren't really in a strong position to forbid it here.

Reply:
In contrast keeping the user from accessing the bundle before diplomatic negotiation has resolved seems very valuable still.